### PR TITLE
Refactor GameWindow to not extend OpenTK.GameWindow

### LINKED ISSUE #1102: Refactor GameWindow to not extend OpenTK.GameWindow
As part of the upcoming iOS/Android builds, we will need to look at refactoring our `GameWindow` class.  It currently extends `OpenTK.GameWindow`, which is primarily for desktop platforms.  Attempting to use it on iOS or Android throws a `NotImplementedException`.

Proposal is to decouple our `GameWindow` class from `OpenTK.GameWindow` and instead use an `OpenTK.IGameWindow` property that's assigned the correct implementation for the platform.  OpenTK will assist us with most of this on the desktop side, but iOS/Android will need to manually instantiate and assign the window (not a big deal).

### PR DESCRIPTION
**Pull request by swoolcock**
Tested on macOS, please test on Windows and Linux.
Closes #1102 

Will require a change to the osu.Desktop project unless we make a wrapper in `GameWindow` for `Icon`, `Title`, and `FileDrop` that redirect to `Implementation`.

Edit: Backported commit: https://github.com/ppy/osu-framework/commit/c7935ae3ade6bd4ba1ce69218f7768841bb18a80

### CODE REVIEW

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 1]
```diff
@@ -137,9 +137,9 @@ private Vector2 getUpdatedPosition(OpenTK.Input.MouseState state, OpenTKMouseSta
                 if (mapAbsoluteInputToWindow)
                 {
                     // map directly to local window
-                    currentPosition.X = ((float)((state.X - raw_input_resolution / 2f) * sensitivity.Value) + raw_input_resolution / 2f) / raw_input_resolution * host.Window.Width;
+                    currentPosition.X = ((float)((state.X - raw_input_resolution / 2f) * sensitivity.Value) + raw_input_resolution / 2f) / raw_input_resolution * host.Window.Implementation.Width;
                     currentPosition.Y = ((float)((state.Y - raw_input_resolution / 2f) * sensitivity.Value) + raw_input_resolution / 2f) / raw_input_resolution
-                                        * host.Window.Height;
+                        * host.Window.Implementation.Height;
```
Comment (smoogipoo) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: Let's put this on the same line as `currentPosition.X` does.

File: osu.Framework/Platform/GameWindow.cs [hunk 2]
```diff
@@ -8,10 +8,11 @@
 using OpenTK;
 using OpenTK.Graphics;
 using OpenTK.Graphics.ES30;
-
+using OpenTK.Platform;
+
 namespace osu.Framework.Platform
 {
-    public abstract class GameWindow : OpenTK.GameWindow
+    public abstract class GameWindow
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameWindow.cs]: Just to reduce the sheer number of `.Implementation` calls around the code, can we make this inherit `IGameWindow` and act as a proxy to `Implementation`? `Implementation` itself should probably be private or protected at most.

File: osu.Framework/Platform/GameWindow.cs [hunk 3]
```diff
@@ -63,8 +61,18 @@ protected GameWindow(int width, int height)
...
+            : this(new OpenTK.GameWindow(width, height, new GraphicsMode(GraphicsMode.Default.ColorFormat, GraphicsMode.Default.Depth, GraphicsMode.Default.Stencil, GraphicsMode.Default.Samples, GraphicsMode.Default.AccumulatorFormat, 3)))
+        {
+            var gw = (OpenTK.GameWindow)Implementation;
+
+            gw.Closing += (sender, e) => e.Cancel = ExitRequested?.Invoke() ?? false;
+            gw.Closed += (sender, e) => Exited?.Invoke();
 
-            Context.MakeCurrent(null);
+            gw.MouseEnter += (sender, args) => CursorInWindow = true;
+            gw.MouseLeave += (sender, args) => CursorInWindow = false;
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameWindow.cs]: It looks like these should be a part of the `INativeWindow` interface, which `IGameWindow` implements. Can we move this to the other ctor?

File: osu.Framework/Platform/Windows/WindowsGameWindow.cs [hunk 4]
```diff
@@ -7,15 +7,15 @@ namespace osu.Framework.Platform.Windows
...
-                return;
-            }
+            Implementation.KeyDown += OnKeyDown;
+        }
 
-            base.OnKeyDown(e);
+        protected void OnKeyDown(object sender, KeyboardKeyEventArgs e)
+        {
+            if (e.Key == Key.F4 && e.Alt)
+                Implementation.Exit();
```
Comment (smoogipoo) [inline osu.Framework/Platform/Windows/WindowsGameWindow.cs]: Along with the previous suggestions, we should be able to fix this potential little bit of unsafety - this code previously early-returned on Alt+F4.

Essentially it'd be something like
```
IGameWindow implementation; // private
implementation.KeyDown += OnKeyDown;

virtual OnKeyDown()
{
    KeyDown?.Invoke();
}
```

Comment (swoolcock) [reply osu.Framework/Platform/Windows/WindowsGameWindow.cs]: Can I get a bit of clarification on this point please?  I'm not sure how your example differs from the code in the PR.

Comment (smoogipoo) [reply osu.Framework/Platform/Windows/WindowsGameWindow.cs]: Then WindowsGameWindow can override it and return after calling `.Exit`, as it does on master, without then invoking any subscribers of the KeyDown event.

**Comment by peppy** [comment]
This looks to have not been tested on windows. Causes silent crash on startup.

**Pull request merged**